bpf/bpf_lxc.c

```diff
@@ -955,6 +1033,66 @@ int tail_ipv4_policy(struct __sk_buff *skb)
 	skb->cb[0] = skb->mark; // essential for proxy ingress, see bpf_ipsec.c
 	return ret;
 }
+
+declare_tailcall_if(__and(is_defined(ENABLE_IPV4), is_defined(ENABLE_IPV6)), CILIUM_CALL_IPV4_TO_ENDPOINT)
+int tail_ipv4_to_endpoint(struct __sk_buff *skb)
+{
+	__u32 src_identity = skb->cb[CB_SRC_LABEL];
+	struct ep_config *cfg = lookup_ep_config();
+	int ret, forwarding_reason;
+	void *data, *data_end;
+	struct iphdr *ip4;
+
+	if (!revalidate_data(skb, &data, &data_end, &ip4)) {
+		ret = DROP_INVALID;
+		goto out;
+	}
+
+	if (cfg == NULL) {
+		ret = DROP_NO_CONFIG;
+		goto out;
+	}
+
+	/* Packets from the proxy will already have a real identity. */
+	if (identity_is_reserved(src_identity)) {
+		struct remote_endpoint_info *info;
+
+		info = ipcache_lookup4(&IPCACHE_MAP, ip4->saddr, V4_CACHE_KEY_LEN);
+		if (info != NULL) {
+			__u32 sec_label = info->sec_label;
+			if (sec_label) {
+				/* When SNAT is enabled on traffic ingressing
+				 * into Cilium, all traffic from the world will
+				 * have a source IP of the host. It will only
+				 * actually be from the host if "src_identity"
+				 * (passed into this function) reports the src
+				 * as the host. So we can ignore the ipcache
+				 * if it reports the source as HOST_ID.
+				 */
+				if (sec_label != HOST_ID)
+					src_identity = sec_label;
+			}
+		}
+		cilium_dbg(skb, info ? DBG_IP_ID_MAP_SUCCEED4 : DBG_IP_ID_MAP_FAILED4,
+			   ip4->saddr, src_identity);
+	}
+
+	cilium_dbg(skb, DBG_LOCAL_DELIVERY, LXC_ID, SECLABEL);
+
+#if defined LOCAL_DELIVERY_METRICS
+	update_metrics(skb->len, METRIC_INGRESS, REASON_FORWARDED);
+#endif
+
+	skb->cb[CB_SRC_LABEL] = 0;
+	ret = ipv4_policy(skb, 0, src_identity, &forwarding_reason, cfg);
```

Comment:
How is L7 proxy handled here? If this is TC egress for veth to a container, it'll return `TC_ACT_OK` and pass through the container instead of redirecting to the proxy, right?

Reply:
I've tested this before tproxy was merged. It is now broken. I'll have to see whether we can redirect to ingress of the host facing device

Reply:
Fixed